File: cli/internal/importers/bitwarden.go
package importers

import (
	"encoding/json"
	"fmt"
)

type bitwardenExport struct {
	Encrypted bool            `json:"encrypted"`
	Items     []bitwardenItem `json:"items"`
}

type bitwardenItem struct {
	Type   int    `json:"type"`
	Name   string `json:"name"`
	SSHKey *struct {
		PrivateKey     string `json:"privateKey"`
		PublicKey      string `json:"publicKey"`
		KeyFingerprint string `json:"keyFingerprint"`
	} `json:"sshKey"`
}

func ParseBitwarden(data []byte) ([]ImportedKey, error) {
	var export bitwardenExport
	if err := json.Unmarshal(data, &export); err != nil {
		return nil, fmt.Errorf("parsing Bitwarden export: %w", err)
	}
	if export.Encrypted {
		return nil, fmt.Errorf("encrypted Bitwarden exports are not supported -- export as unencrypted JSON")
	}

	var keys []ImportedKey
	for _, item := range export.Items {
		if item.Type != 5 || item.SSHKey == nil || item.SSHKey.PrivateKey == "" {
			continue
		}
		keys = append(keys, ImportedKey{
			Name:       SanitizeName(item.Name),
			PrivateKey: item.SSHKey.PrivateKey,
			PublicKey:  item.SSHKey.PublicKey,
		})
	}
	return keys, nil
}

File: web/src/lib/importers/bitwarden.ts
import { DEFAULT_IMPORTED_NAME, normalizeImportedName } from "./name";
import type { ImportedKey } from "./types";

export function parseBitwarden(data: string): ImportedKey[] {
	const parsed = JSON.parse(data);
	if (parsed.encrypted) throw new Error("Encrypted Bitwarden exports are not supported. Export as unencrypted JSON.");

	const keys: ImportedKey[] = [];
	for (const item of parsed.items || []) {
		if (item.type !== 5 || !item.sshKey?.privateKey) continue;
		keys.push({
			name: normalizeImportedName(item.name || DEFAULT_IMPORTED_NAME),
			privateKey: item.sshKey.privateKey,
			publicKey: item.sshKey.publicKey,
			fingerprint: item.sshKey.keyFingerprint,
		});
	}
	return keys;
}
